contract-dev/vanity.mdx

```diff
@@ -1,7 +1,176 @@
 ---
-title: "Vanity"
+title: "Using vanity contract"
 ---
 
-import { Stub } from '/snippets/stub.jsx';
+A [vanity contract](https://github.com/ton-community/vanity-contract) allows customizing the address of a smart contract being deployed. It does this by making its own `StateInit` depend on constant data that is randomly generated many times until a desired address is found. It is often used to deploy contracts with a specific prefix or suffix so it is visible in block explorers.
 
-<Stub issue="385" />
+The contract code and data are included in the vanity deploy message. The vanity contract is first deployed with a `StateInit` that produces the desired address (see [Addresses overview](/foundations/addresses/overview#account-id)), then immediately sets its actual state from the payload.
+
+### Prerequisites
+
+- Python 3 and `pyopencl`
+- OpenCL‑capable GPU
+- [Blueprint](/contract-dev/blueprint/overview)
+
+## How it works
+
+The vanity contract code:
+
+```func
+(int) slice_equal(slice s1, slice s2) asm "SDEQ";
+
+() recv_internal(cell in_msg_cell, slice in_msg) impure {
+    ;; Parse data
+    var ds = get_data().begin_parse();
+    ds~skip_bits(5); ;; Padding
+    var owner = ds~load_msg_addr();
+    ds~skip_bits(256);
+    ds.end_parse();
+
+    ;; Parse message
+    var cs = in_msg_cell.begin_parse();
+    var flags = cs~load_uint(4);  ;; int_msg_info$0 ihr_disabled:Bool bounce:Bool bounced:Bool
+    slice sender = cs~load_msg_addr();
+
+    ;; Allow deployment only to owner
+    throw_unless(8, slice_equal(sender, owner));
+
+    ;; Set code and data
+    var code = in_msg~load_ref();
+    var data = in_msg~load_ref();
+    in_msg.end_parse();
+    set_code(code);
+    set_data(data);
+}
+```
+
+It checks whether the message comes from the owner specified in data, then replaces its code and data with the ones provided in the incoming message. The 256‑bit salt is stored in this contract’s `StateInit` in addition to five padding bits and the owner address. The salt is not used by vanity's code and is skipped in the contract code with `ds~skip_bits(256);`, but it determines the resulting address.
+
+Since contract addresses depend on the hash of `StateInit`, adding data to `StateInit` changes the address deterministically. The vanity contract uses this by generating many random salt values included in `StateInit` and searching for ones that produce the desired address. This is analogous to proof‑of‑work nonce search. The salt generation logic lives in a separate Python script that uses OpenCL for faster search on a GPU.
```

Comment:
> The vanity contract uses this by generating many random salt values

Let's make more explicit the fact that this search is done offchain, not in the contract code itself.